src/honey_badger.rs

```diff
@@ -72,30 +75,42 @@ where
     type Message = Message<NodeUid>;
     type Error = Error;
 
-    fn input(&mut self, input: Self::Input) -> HoneyBadgerResult<()> {
-        self.add_transactions(iter::once(input))
+    fn input(&mut self, input: Self::Input) -> HoneyBadgerResult<FaultLog<NodeUid>> {
+        self.add_transactions(iter::once(input))?;
+        match self.fault_log.get(&self.epoch) {
+            Some(fault_log) => Ok(fault_log.clone()),
+            None => Ok(FaultLog::new()),
+        }
     }
 
     fn handle_message(
         &mut self,
         sender_id: &NodeUid,
         message: Self::Message,
-    ) -> HoneyBadgerResult<()> {
+    ) -> HoneyBadgerResult<FaultLog<NodeUid>> {
         if !self.netinfo.all_uids().contains(sender_id) {
             return Err(ErrorKind::UnknownSender.into());
         }
         let Message { epoch, content } = message;
         if epoch < self.epoch {
-            // Ignore all messages from past epochs.
-            return Ok(());
+            match self.fault_log.get(&self.epoch) {
+                Some(fault_log) => return Ok(fault_log.clone()),
+                None => return Ok(FaultLog::new()),
+            };
+
+
         }
         match content {
             MessageContent::CommonSubset(cs_msg) => {
-                self.handle_common_subset_message(sender_id, epoch, cs_msg)
+                self.handle_common_subset_message(sender_id, epoch, cs_msg)?
             }
             MessageContent::DecryptionShare { proposer_id, share } => {
-                self.handle_decryption_share_message(sender_id, epoch, proposer_id, share)
+                self.handle_decryption_share_message(sender_id, epoch, proposer_id, share)?
             }
+        };
+        match self.fault_log.get(&self.epoch) {
```

Comment:
The returned value will keep on growing. Does it make sense to `clone` the growing log after every received message? The `HoneyBadger` epoch can be dropped from the logs for now. When a node becomes faulty, the client should take any necessary actions immediately, so the epoch information is redundant. If you don't use epochs, `HoneyBadger` can return the logs instead of saving them locally. Returning the logs and not saving them solves the problem with the growing epoch log.

Reply:
Yea, I think that I'm going to remove the `fault_log` field from HoneyBadger and DynamicHoneyBadger.